doc/man/man4/tali.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/tali.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH TALI 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B tali
\- Tekelec's Adaptation Layer Interface
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/mtpi.h>
.br
.B #include <sys/mtpi_ioctl.h>
.br
.B #include <sys/tali_ioctl.h>
.HP
.BI "int " tali_stream " = open(" device ", " flags );
.HP
.BI "int " tali_xtistr " = t_open(" device ", " flags );
.HP
.BI "int " tali_socket " = socket(" \fRPF_SS7\fI ", " socket_type ", " \fRSS7PROTO_TALI\fI );
.\"
.\"
.SH DESCRIPTION
.PP
.B TALI
is an implementation of the Tekelec Adaptation Layer Interface (TALI) defined in
RFC 3094.
.[
rfc3094
.]
.B TALI
provides a back haul of the services of the
.BR mtp (4)
over an IP network using Transmission Control Protocol,
.BR tcp (4)
as defined in RFC 793.
.[
rfc793
.]
Also, it provides a building block for redundant and distributed SS7
systems.  (See
.IR \(lqNOTICES\(rq ,
below.)
.PP
.B TALI
is implemented as a
.IR STREAMS (4)
multiplexing driver.
.BR mtp "(4) or " tcp (4)
streams are linked under the
.BR TALI
multiplexing driver using the
.IR I_LINK "(7) or " I_PLINK (7)
commands of
.BR streamio (7),
to provide transport or direct MTP services
.[
q704
.]
.[
mtpv1
.]
.[
t1.111
.]
.[
jtq704
.]
to
.BR TALI .
Once linked, the streams must be configured using
.BR tali_ioctl (4).
.PP
.B TALI
provides the following stream types that are directly compatible with the
stream types provided by
.BR mtp (4).
To select one of these stream types,
.I device
can be one of the following:
.TP
.I /dev/tali
This device provides the
.B MTP
protocol service interface using the M-primitives of the Message Transfer Part
Interface described in
.BR mtpi (7).
This interface is suitable for use with other SS7 modules and drivers.
.TP
.IR /dev/tali-npi-cl ", " /dev/tali-npi-co
This devices provides the
.BR MTP
protocol service interface using the N-primitives of the Network Provider
Interface,
.[
NPI Revision 2.0.0
.]
.BR npi (7),
used for the Message Transfer Part as described in
.BR npi_mtp (7).
The
.BR npi (7)
interface provided on a stream opening this device is suitable for use with
the
.BR sockets (3)
library
.RB socket (3)
call
and provides the
.B MTP
interface using sockets as described in
.BR mtp_socket (7).
.I /dev/tali-npi-cl
provides a connectionless
.B N_CLNS
network service provider;
.I /dev/tali-npi-co
provides a connection-oriented
.B N_CONS
network service provider.
.TP
.IR /dev/tali-tpi-cl ", " /dev/tali-tpi-co
This device provides the
.B MTP
protocol service interface using the T-primitives of the
Transport Provider Interface,
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
user for the Message Transfer Part as described in
.BR tpi_mtp (7).
The
.BR tpi (7)
interface provided on a stream opening this device is suitable for use with
the
.BR xnet (3)
or
.BR xti (3)
library
.BR t_open (3)
call and provides the
.B MTP
interface using XTI
.[
xti revision 1.0
.]
.[
xns99
.]
as described in
.BR xti_mtp (3).
.I /dev/tali-tpi-cl
provides a connectionless
.B T_CLNS
transport service provider;
.I /dev/tali-tpi-co
provides a pseudo-connection-oriented
.B T_CONS
transport service provider.
.PP
.B TALI
provides two socket types used by the
.BR sockets (3)
library
.BR socket (3)
call.
.B TALI
provides sockets in the protocol family
.B PF_SS7
using the protocol number
.BR SS7PROTO_TALI .
The
.I socket_type
argument can be one of the following:
.TP
.B SOCK_RDM
Provides a connectionless datagram service with reliable delivery including
retransmission.  Message boundaaries are preserved.
This socket uses the underlying
.I /dev/tali-tpi-cl
transport service provider.
.TP
.B SOCK_SEQPACKET
Provides a pseudo-connection-oriented packet service with reliable delivery
including retransmission.  Message boundaries are preserved.  Messages can be
sent and received in order.
This socket uses the underlying
.I /dev/tali-tpi-co
transport service provider.
.PP
The following entries in
.B /etc/sock2path
provide the information necessary for
.BR socksys (4)
to use these devices:
.sp
.TS
tab(:);
llllll.
PF_SS7:SOCK_RDM:SS7PROTO_TALI:/dev/tali-tpi-cl:tpi:clts
PF_SS7:SOCK_SEQPACKET:SS7PROTO_TALI:/dev/tali-tpi-co:tpi:cots
.TE
.PP
For additional information, see
.BR sock2path (5).
.\"
.\"
.SH NOTICES
.PP
.B TALI
can be configured on a system as a drop in transparent replacement for
.BR mtp (4).
In such a configuration, users opening streams or sockets as defined in
.BR mtp (4)
will be opening
.B tali
streams and sockets, but will be unaware of the difference.  In
.BR OpenSS7 ,
.B tali
is used as the primary mechanism of redundancy and distribution at the
.B MTP
level, and, when properly configured (see
.BR ss7confd (8)),
.B tali
functions transparently to the MTP-User.
.PP
Also, opening and binding a
.B tali

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a

src/drivers/tali.c
/*****************************************************************************

 @(#) File: src/drivers/tali.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/tali.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#include <sys/os7/compat.h>

#include "tali.h"
#include "tali_data.h"

/*
 *  TALI MULTIPLEXING DRIVER
 *  -------------------------------------------------------------------------
 *  This is a multiplexing driver to TALI.  When TALI is opened by its control
 *  stream (typically a tali configuration daemon), it provides a control
 *  channel for configuration and routing.  When TALI is opened by one of its
 *  user streams, it provides either a normal OpenSS7 MTP, ISUP, or SCCP stream.
 */

#define TALI_DESCRIP	"TALI STREAMS Multiplexing Driver"
#define TALI_EXTRA	"Part of the OpenSS7 SS7 Stack for Linux Fast-STREAMS"
#define TALI_REVISION	"OpenSS7 src/drivers/tali.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define TALI_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define TALI_DEVICE	"Part of the OpenSS7 Stack for Linux Fast STREAMS."
#define TALI_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define TALI_LICENSE	"GPL"
#define TALI_BANNER	TALI_DESCRIP	"\n" \
			TALI_EXTRA	"\n" \
			TALI_REVISION	"\n" \
			TALI_COPYRIGHT	"\n" \
			TALI_DEVICE	"\n" \
			TALI_CONTACT	"\n"
#define TALI_SPLASH	TALI_DEVICE	" - " \
			TALI_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(TALI_CONTACT);
MODULE_DESCRIPTION(TALI_DESCRIP);
MODULE_SUPPORTED_DEVICE(TALI_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(TALI_LICENSE);
#endif
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-tali");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#define TALI_DRV_ID	CONFIG_STREAMS_TALI_MODID
#define TALI_DRV_NAME	CONFIG_STREAMS_TALI_NAME
#define TALI_CMAJORS	CONFIG_STREAMS_TALI_NMAJORS
#define TALI_CMAJOR_0	CONFIG_STREAMS_TALI_MAJOR
#define TALI_UNITS	CONFIG_STREAMS_TALI_UNITS

#define TALI_MTP_CMINOR		1
#define TALI_SCCP_CMINOR	2
#define TALI_ISUP_CMINOR	3
#define TALI_FREE_CMINOR	4

/*
 *  =========================================================================
 *
 *  STREAM Definitions
 *
 *  =========================================================================
 */

#define DRV_ID		TALI_DRV_ID
#define DRV_NAME	TALI_DRV_NAME
#define CMAJORS		TALI_CMAJORS
#define CMAJOR_0	TALI_CMAJOR_0
#define UNITS		TALI_UNITS
#ifdef MODULE
#define DRV_BANNER	TALI_BANNER
#else				/* MODULE */
#define DRV_BANNER	TALI_SPLASH
#endif				/* MODULE */

STATIC struct module_info tali_minfo = {
	.mi_idnum = DRV_ID,		/* Module ID number */
	.mi_idname = DRV_NAME,		/* Module name */
	.mi_minpsz = 0,			/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,		/* Max packet size accepted */
	.mi_hiwat = 1,			/* Hi water mark */
	.mi_lowat = 0,			/* Lo water mark */
};

STATIC struct module_stat tali_mstat = { 0, };

STATIC int tali_open(queue_t *, dev_t *, int, int, cred_t *);
STATIC int tali_close(queue_t *, int, cred_t *);

STATIC int tali_rdprim(queue_t *, mblk_t *);
STATIC int tali_wrprim(queue_t *, mblk_t *);

STATIC int mtp_rdprim(queue_t *, mblk_t *);
STATIC int mtp_wrprim(queue_t *, mblk_t *);

STATIC int sccp_rdprim(queue_t *, mblk_t *);
STATIC int sccp_wrprim(queue_t *, mblk_t *);

STATIC int isup_rdprim(queue_t *, mblk_t *);
STATIC int isup_wrprim(queue_t *, mblk_t *);

STATIC struct qinit tali_rdinit {
	.qi_putp = ss7_oput,		/* Read put (msg from below) */
	.qi_srvp = ss7_osrv,		/* Read queue service */
	.qi_qopen = tali_open,		/* Each open */
	.qi_qclose = tali_close,	/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &tali_minfo,	/* Information */
	.qi_mstat = &tali_mstat,	/* Statistics */
};

STATIC struct qinit tali_wrinit {
	.qi_putp = ss7_iput,		/* Write put (msg from above) */
	.qi_srvp = ss7_isrv,		/* Write queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &tali_minfo,	/* Information */
	.qi_mstat = &tali_mstat,	/* Statistics */
};

STATIC int sg_muxrprim(queue_t *, mblk_t *);
STATIC int sg_muxwprim(queue_t *, mblk_t *);

STATIC int as_muxrprim(queue_t *, mblk_t *);
STATIC int as_muxwprim(queue_t *, mblk_t *);

STATIC int mtp_muxrprim(queue_t *, mblk_t *);
STATIC int mtp_muxwprim(queue_t *, mblk_t *);

STATIC int sccp_muxrprim(queue_t *, mblk_t *);
STATIC int sccp_muxwprim(queue_t *, mblk_t *);

STATIC int isup_muxrprim(queue_t *, mblk_t *);
STATIC int isup_muxwprim(queue_t *, mblk_t *);

STATIC struct qinit tali_muxrinit {
	.qi_putp = ss7_iput,		/* Read put (msg from below) */
	.qi_srvp = ss7_isrv,		/* Read queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &tali_minfo,	/* Information */
	.qi_mstat = &tali_mstat,	/* Statistics */
};

STATIC struct qinit tali_muxwinit {
	.qi_putp = ss7_oput,		/* Write put (msg from above) */
	.qi_srvp = ss7_osrv,		/* Write queue service */
	.qi_qopen = NULL,		/* Each open */
	.qi_qclose = NULL,		/* Last close */
	.qi_qadmin = NULL,		/* Admin (not used) */
	.qi_minfo = &tali_minfo,	/* Information */
	.qi_mstat = &tali_mstat,	/* Statistics */
};

MODULE_STATIC struct streamtab tali_info = {
	.st_rdinit = &tali_rdinit,	/* Upper read queue */
	.st_wrinit = &tali_wrinit,	/* Upper write queue */
	.st_muxrinit = &tali_muxrinit,	/* Lower read queue */
	.st_muxwinit = &tali_muxwinit,	/* Lower write queue */
};

/*
 *  =========================================================================
 *
 *  Private Structures
 *
 *  =========================================================================
 */

/*
 *  =========================================================================
 *
 *  Cache initialization and termination.
 *
 *  =========================================================================
 */

extern kmem_cachep_t tali_pp_cachep = NULL;
extern kmem_cachep_t tali_xp_cachep = NULL;
extern kmem_cachep_t tali_gp_cachep = NULL;
extern kmem_cachep_t tali_as_cachep = NULL;
extern kmem_cachep_t tali_ap_cachep = NULL;
extern kmem_cachep_t tali_sp_cachep = NULL;
extern kmem_cachep_t tali_np_cachep = NULL;

STATIC tali_pp_cache_allocated = 0;
STATIC tali_xp_cache_allocated = 0;
STATIC tali_gp_cache_allocated = 0;
STATIC tali_as_cache_allocated = 0;
STATIC tali_ap_cache_allocated = 0;
STATIC tali_sp_cache_allocated = 0;
STATIC tali_np_cache_allocated = 0;

STATIC void
tali_term_caches(void)
{
	if (tali_pp_cachep) {
		if (tali_pp_cache_allocated)
			kmem_cache_destroy(tali_pp_cachep);
		tali_pp_cachep = NULL;
	}
	if (tali_xp_cachep) {
		if (tali_xp_cache_allocated)
			kmem_cache_destroy(tali_xp_cachep);
		tali_xp_cachep = NULL;
	}
	if (tali_gp_cachep) {
		if (tali_gp_cache_allocated)
			kmem_cache_destroy(tali_gp_cachep);
		tali_gp_cachep = NULL;
	}
	if (tali_sp_cachep) {
		if (tali_sp_cache_allocated)
			kmem_cache_destroy(tali_sp_cachep);
		tali_sp_cachep = NULL;
	}
	if (tali_as_cachep) {
		if (tali_as_cache_allocated)
			kmem_cache_destroy(tali_as_cachep);
		tali_as_cachep = NULL;
	}
	if (tali_ap_cachep) {
		if (tali_ap_cache_allocated)
			kmem_cache_destroy(tali_ap_cachep);
		tali_ap_cachep = NULL;
	}
	if (tali_np_cachep) {
		if (tali_np_cache_allocated)
			kmem_cache_destroy(tali_np_cachep);
		tali_np_cachep = NULL;
	}
	return;
}

STATIC int
tali_init_caches(void)
{
	if (!(tali_pp_cachep))
		if (!(tali_pp_cachep = kmem_create_cache("tali_pp_cachep", sizeof(pp_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_pp_cache_allocated = 1;
	if (!(tali_xp_cachep))
		if (!(tali_xp_cachep = kmem_create_cache("tali_xp_cachep", sizeof(xp_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_xp_cache_allocated = 1;
	if (!(tali_gp_cachep))
		if (!(tali_gp_cachep = kmem_create_cache("tali_gp_cachep", sizeof(gp_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_gp_cache_allocated = 1;
	if (!(tali_sp_cachep))
		if (!(tali_sp_cachep = kmem_create_cache("tali_sp_cachep", sizeof(sp_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_sp_cache_allocated = 1;
	if (!(tali_as_cachep))
		if (!(tali_as_cachep = kmem_create_cache("tali_as_cachep", sizeof(as_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_as_cache_allocated = 1;
	if (!(tali_ap_cachep))
		if (!(tali_ap_cachep = kmem_create_cache("tali_ap_cachep", sizeof(ap_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_ap_cache_allocated = 1;
	if (!(tali_np_cachep))
		if (!(tali_np_cachep = kmem_create_cache("tali_np_cachep", sizeof(np_t),
							 0, SLAB_HWCACHE_ALIGN, NULL, NULL)))
			goto tali_init_caches_failed;
		else
			tali_np_cache_allocated = 1;
	return (0);

      tali_init_caches_failed:
	tali_term_caches();
	return (ENOMEM);
}

/*
 *  =========================================================================
 *
 *  OPEN and CLOSE
 *
 *  =========================================================================
 */
STATIC major_t tali_majors[CMAJORS] = { CMAJOR_0, };

/*
 *  OPEN
 *  -------------------------------------------------------------------------
 */
STATIC int
tali_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *crp)
{
	psw_t flags;
	int mindex = 0;
	major_t cmajor = bmajor = getmajor(*devp);
	minor_t cminor = bminor = getminor(*devp);
	struct tali *ta, **tpp;

	MOD_INC_USE_COUNT;
	if (q->q_ptr != NULL) {
		MOD_DEC_USE_COUNT;
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q) - q_next) {
		ptrace(("%s: ERROR: cannot push as module\n", DRV_NAME));
		MOD_DEC_USE_COUNT;
		return (EIO);
	}
	if (cmajor != CMAJOR_0 || cminor > TALI_ISUP_CMINOR) {
		MOD_DEC_USE_COUNT;
		return (ENXIO);
	}
	/* allocate a new device */
	cminor = TALI_FREE_CMINOR;
	spin_lock_irqsave(&master.lock, flags);
	for (tpp = &master.ta.list; *tpp; tpp = &(*tpp)->next) {
		major_t dmajor = (*tpp)->u.dev.cmajor;

		if (cmajor != dmajor) {
			minor_t dminor = (*tpp)->u.dev.cminor;

			if (cminor < dminor)
				break;
			if (cminor > dminor)
				continue;
			if (cminor == dminor) {
				if (++cminor >= NMINORS) {
					if (++mindex >= CMAJORS || !(cmajor = tali_majors[mindex]))
						break;
					cminor = 0;
				}
				continue;
			}
		}
	}
	if (mindex >= CMAJORS || !cmajor) {
		ptrace(("%s: ERROR: no device numbers available\n", DRV_NAME));
		spin_unlock_irqrestore(&master.lock, flags);
		MOD_DEC_USE_COUNT;
		return (ENXIO);
	}
	printd(("%s: opened character device %hu:%hu\n", DRV_NAME, cmajor, cminor));
	*devp = makedevice(cmajor, cminor);
	if (!(ta = tali_alloc_ta(q, tpp, cmajor, cminor, crp, bminor))) {
		ptrace(("%s: ERROR: no memory\n", DRV_NAME));
		spin_unlock_irqrestore(&master.lock, flags);
		MOD_DEC_USE_COUNT;
		return (ENOMEM);
	}
	spin_unlock_irqrestore(&master.lock, flags);
	return (0);
}

/*
 *  CLOSE
 *  -------------------------------------------------------------------------
 */
STATIC int
tali_close(queue_t *q, int flag, cred_t *crp)
{
	struct tali *ta = TA_PRIV(q);
	psw_t flags;

	(void) ta;
	printd(("%s: %p: closing character device %hu:%hu\n", DRV_NAME, ta, ta->u.dev.cmajor,
		ta->u.dev.cminor));
	spin_lock_irqsave(&master.lock, flags);
	{
		tali_free_ta(q);
	}
	spin_unlock_irqrestore(&master.lock, flags);
	MOD_DEC_USE_COUNT;
	return (0);
}

/*
 *  =========================================================================
 *
 *  Registration and initialization
 *
 *  =========================================================================
 */
#ifdef LINUX
/*
 *  Linux Registration
 *  -------------------------------------------------------------------------
 */

unsigned short modid = DRV_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for the TALI driver. (0 for allocation.)");

major_t major = CMAJOR_0;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Device number for the TALI driver. (0 for allocation.)");

/*
 *  Linux Fast-STREAMS Registration
 *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 */
STATIC struct cdevsw tali_cdev = {
	.d_name = DRV_NAME,
	.d_str = &taliinfo,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR,
	.d_kmod = THIS_MODULE,
};

STATIC int
tali_register_strdev(major_t major)
{
	int err;

	if ((err = register_strdev(&tali_cdev, major)) < 0)
		return (err);
	return (0);
}

STATIC int
tali_unregister_strdev(major_t major)
{
	int err;

	if ((err = unregister_strdev(&tali_cdev, major)) < 0)
		return (err);
	return (0);
}

MODULE_STATIC void __exit
taliterminate(void)
{
	int err, mindex;

	for (mindex = CMAJORS - 1; mindex >= 0; mindex--) {
		if (tali_majors[mindex]) {
			if ((err = tali_unregister_strdev(tali_majors[mindex])))
				cmn_err(CE_PANIC, "%s: cannot unregister major %d", DRV_NAME,
					tali_majors[mindex]);
			if (mindex)
				tali_majors[mindex] = 0;
		}
	}
	if ((err = tali_term_caches()))
		cmn_err(CE_WARN, "%s: could not terminate caches", DRV_NAME);
	return;
}

MODULE_STATIC int __init
taliinit(void)
{
	int err, mindex = 0;

	cmn_err(CE_NOTE, DRV_BANNER);	/* console splash */
	if ((err = tali_init_caches())) {
		cmn_err(CE_WARN, "%s: could not init caches, err = %d", DRV_NAME, err);
		taliterminate();
		return (err);
	}
	for (mindex = 0; mindex < CMAJORS; mindex++) {
		if ((err = tali_register_strdev(tali_majors[mindex])) < 0) {
			if (mindex) {
				cmn_err(CE_WARN, "%s: could not register major %d", DRV_NAME,
					tali_majors[mindex]);
				continue;
			} else {
				cmn_err(CE_WARN, "%s: could not register driver, err = %d",
					DRV_NAME, err);
				taliterminate();
				return (err);
			}
		}
		if (tali_majors[mindex] == 0)
			tali_majors[mindex] = err;
#if 0
		LIS_DEVFLAGS(tali_majors[mindex]) |= LIS_MODFLG_CLONE;
#endif
		if (major == 0)
			major = tali_majors[0];
	}
	return (0);
}

/*
 *  Linux Kernel Module Initialization
 *  -------------------------------------------------------------------------
 */
module_init(taliinit);
module_exit(taliterminate);

#endif				/* LINUX */
